src/System.Net.Security/src/System/Net/NegotiateStreamPal.Unix.cs

```diff
@@ -146,8 +250,66 @@ internal static Exception CreateExceptionFromError(SecurityStatusPal statusCode)
             out int newOffset,
             uint sequenceNumber)
         {
-            throw new PlatformNotSupportedException();
+            throw new NotImplementedException("DecryptNtlm: NTLM is not yet supported");
         }
 
+        private static SecurityStatusPal EstablishSecurityContext(
+          SafeFreeNegoCredentials credential,
+          ref SafeDeleteContext context,
+          bool isNtlm,
+          string targetName,
+          ContextFlagsPal inFlags,
+          SecurityBuffer inputBuffer,
+          SecurityBuffer outputBuffer,
+          ref ContextFlagsPal outFlags)
+        {
+            Debug.Assert(!isNtlm, "EstablishSecurityContext: NTLM is not yet supported");
+
+            if (context == null)
+            {
+                context = new SafeDeleteNegoContext(credential, targetName);
+            }
+
+            SafeDeleteNegoContext negoContext = (SafeDeleteNegoContext)context;
+            try
+            {
+                Interop.NetSecurityNative.GssFlags inputFlags = ContextFlagsAdapterPal.GetInteropFromContextFlagsPal(inFlags);
+                uint outputFlags;
+                SafeGssContextHandle contextHandle = negoContext.GssContext;
+                bool done = Interop.GssApi.EstablishSecurityContext(
+                                  ref contextHandle,
+                                  credential.GssCredential,
+                                  isNtlm,
+                                  negoContext.TargetName,
+                                  inputFlags,
+                                  ((inputBuffer != null) ? inputBuffer.token : null),
+                                  out outputBuffer.token,
+                                  out outputFlags);
+
+                Debug.Assert(outputBuffer.token != null, "Unexpected null buffer returned by GssApi");
+                outputBuffer.size = outputBuffer.token.Length;
+                outputBuffer.offset = 0;
+
+                outFlags = ContextFlagsAdapterPal.GetContextFlagsPalFromInterop((Interop.NetSecurityNative.GssFlags)outputFlags);
+
+                // Save the inner context handle for further calls to NetSecurity
+                if (null == negoContext.GssContext)
+                {
+                    negoContext.SetGssContext(contextHandle);
+                }
```

Comment:
What if negoContext.GssContext isn't null?  Must it equal contextHandle, and if so, can we assert that?  Or if it may not be equal, what happens to contextHandle?